src/nyc_trees/apps/survey/management/commands/send_reminder_emails.py

```diff
@@ -60,25 +66,50 @@ def handle(self, *args, **options):
                                  canceled_at__isnull=True,
                                  expires_at__lte=soon))
 
-        user_reservations = {}
-
-        for reservation in expiring_soon:
-            user_data = user_reservations.get(reservation.user_id, [])
-            user_data.append(reservation)
-            user_reservations[reservation.user_id] = user_data
-
-        if not user_reservations:
+        if not expiring_soon:
             self._log("no emails sent")
             return
 
-        for user_id, reservations in user_reservations.items():
-            try:
-                send_reservation_reminder(user_id, reservations=reservations)
-            except (smtplib.SMTPException, SESError):
-                continue
-            self._log("Email sent to user_id: '%s'" % user_id)
-            ids = map(operator.attrgetter('id'), reservations)
-            (BlockfaceReservation
-             .objects
-             .filter(id__in=ids)
-             .update(reminder_sent_at=now()))
+        users = {}
+        for reservation in expiring_soon:
+            user_data = users.get(reservation.user_id, {})
+            user_reservations = user_data.get(_RESERVATIONS, [])
+            user_reservations.append(reservation)
+            user_data[_RESERVATIONS] = user_reservations
+            user_data[_REMAINING_RETRIES] = _RETRY_COUNT
+            users[reservation.user_id] = user_data
+
+        remaining_retries = True
+        while remaining_retries:
+            remaining_retries = False
+            for user_id, user_data in users.items():
+                user_remaining_retries = user_data[_REMAINING_RETRIES]
+
+                if user_remaining_retries == 0:
+                    continue
+
+                reservations = user_data[_RESERVATIONS]
+
+                try:
+                    send_reservation_reminder(
+                        user_id, reservations=reservations)
+                except (smtplib.SMTPException, SESError):
+                    user_remaining_retries -= 1
+                    user_data[_REMAINING_RETRIES] = user_remaining_retries
+                    if user_remaining_retries > 0:
+                        remaining_retries = True
+                    continue
+                finally:
+                    # sleep to avoid hitting request-per-second limits in AWS
+                    # given that we have plenty of time for this tasks to run
+                    time.sleep(0.5)
+
+                self._log("Email sent to user_id: '%s'" % user_id)
+                ids = map(operator.attrgetter('id'), reservations)
+                (BlockfaceReservation
+                 .objects
+                 .filter(id__in=ids)
+                 .update(reminder_sent_at=now()))
```

Comment:
I am just now realizing that management tasks run in an implicit transaction. We could send 1000 emails, mark them as sent, and then, if the process crashes, the transaction will roll back and we will resend all the emails the next day. If we don't notice the crash and it is persistent, we will start spamming people day after day.

If we stick with a bulk transaction, we are going to need to know, immediately, if the transaction rolls back for some reason. It may be less nerve wracking to fire off a task for each email.